lib/Dialect/QUIR/Transforms/MergeCircuits.cpp

```diff
@@ -259,9 +293,119 @@ MergeCircuitsPass::getCircuitOp(CallCircuitOp callCircuitOp,
   return circuitOp;
 }
 
+int MergeCircuitsPass::addArguments(
+    Operation *op, llvm::SmallVector<Value> &callInputValues,
+    llvm::SmallVector<int> &insertedArguments,
+    std::unordered_map<int, int> &reusedArguments, int baseIndex) {
+  int index = baseIndex;
+  for (auto inputValue : op->getOperands()) {
+    auto *search = find(callInputValues, inputValue);
```

Comment:
I'm wondering if it makes sense to use a map or set here instead of vectors. This algorithm is essentially n^2 and I think as the number of qubits on the devices grows this kind of things will become prohibitive.

Reply:
@mbhealy I have re-worked this PR with [e254334](https://github.com/openqasm/qe-compiler/pull/239/commits/e2543340b05412aa15850d837bb3f0dd13e85d88) and [5cf4eca](https://github.com/openqasm/qe-compiler/pull/239/commits/5cf4eca764f6decdb042e0be27ec72835fb26df9) to use and std::unordered_map to cache operand / argument indexes so that the vector does not have to be searched each time.